[pybind] Add pybind API reference docs

larryliu0820 opened the pull request:
[pybind] Add pybind API reference docs

Summary: As titled

Test Plan: https://docs-preview.pytorch.org/pytorch/executorch/6276/runtime-python-api-reference.html

Reviewers:

Subscribers:

Tasks:

Tags:

kirklandsign commented:
Sorry not sure why the bot is attacking this PR

dbort commented on docs/source/runtime-pybind-api-reference.rst:
```diff
@@ -0,0 +1,18 @@
+Runtime Pybind API Reference
```
To let people find this doc I think you'll need to add an entry to this section: https://github.com/pytorch/executorch/blob/2c8b14c5577d63b56e514faa291592938e0d39b3/docs/source/index.rst?plain=1#L126

dbort commented on docs/source/index.rst:
```diff
@@ -131,6 +131,7 @@ Topics in this section will help you get started with ExecuTorch.
    export-to-executorch-api-reference
    executorch-runtime-api-reference
    api-life-cycle
+   runtime-python-api-refernece
```
Suggest reordering to keep all of the API refs first, then the lifecycle doc last.

Also, the file name was misspelled ("referNEce"), which might be why the docs didn't update for this latest version.

```suggestion
   export-to-executorch-api-reference
   executorch-runtime-api-reference
   runtime-python-api-reference
   api-life-cycle
```

dbort commented on runtime/__init__.py:
```diff
@@ -7,7 +7,8 @@
 Example usage:
-.. code-block:: text
+
+.. code-block:: python
```
Just noticed that this example probably won't work as-is since it doesn't import Program or Method. Please try pasting it into a python interpreter to make sure it works

dbort commented on docs/source/runtime-python-api-reference.rst:
```diff
@@ -0,0 +1,17 @@
+ExecuTorch Runtime Python API Reference
+----------------------------------
+The Python ``executorch.runtime`` module wraps the C++ ExecuTorch runtime. It can load and execute serialized ``.pte`` program files: see the `Export to ExecuTorch Tutorial <tutorials/export-to-executorch-tutorial.html>`__ for how to convert a PyTorch ``nn.Module`` to an ExecuTorch ``.pte`` program file. Execution accepts and returns ``torch.Tensor`` values, making it a quick way to validate the  ...
+For detailed information on how APIs evolve and the deprecation process, please refer to the `ExecuTorch API Life Cycle and Deprecation Policy <api-life-cycle.html>`__.
```
Right now this flows into the preceding paragraph, even though it's a separate topic. Would be good to add a blank line to separate it.

```suggestion

For detailed information on how APIs evolve and the deprecation process, please refer to the `ExecuTorch API Life Cycle and Deprecation Policy <api-life-cycle.html>`__.
```

dbort commented on docs/source/runtime-python-api-reference.rst:
```diff
@@ -0,0 +1,17 @@
+ExecuTorch Runtime Python API Reference
+----------------------------------
+The Python ``executorch.runtime`` module wraps the C++ ExecuTorch runtime. It can load and execute serialized ``.pte`` program files: see the `Export to ExecuTorch Tutorial <tutorials/export-to-executorch-tutorial.html>`__ for how to convert a PyTorch ``nn.Module`` to an ExecuTorch ``.pte`` program file. Execution accepts and returns ``torch.Tensor`` values, making it a quick way to validate the  ...
```
Looking at the rendered web page, I see

> The Python executorch.runtime module wraps the C++ ExecuTorch runtime. It can load and execute serialized .pte program files: see the [Export to ExecuTorch Tutorial](https://docs-preview.pytorch.org/pytorch/executorch/6276/tutorials/export-to-executorch-tutorial.html) for how to convert a PyTorch nn.Module to an ExecuTorch .pte program file. Execution accepts and returns torch.Tensor values, making it a quick way to validate the correctness of the program. For detailed information on how APIs evolve and the deprecation process, please refer to the [ExecuTorch API Life Cycle and Deprecation Policy](https://docs-preview.pytorch.org/pytorch/executorch/6276/api-life-cycle.html).
>
> Interface to the native C++ ExecuTorch runtime.

Since the .rst text and this doc comment flow together, it might make sense to move more of the details into the python file's doc comment. Or, add a section like `## executorch.runtime Module` in this file before the `automodule::` to make this separation a little more natural. Right now it's a little jarring.

larryliu0820 committed:
Add pybind API reference docs (#6276)

Summary:
[pybind] Add pybind API reference docs

 As titled


Test Plan: https://docs-preview.pytorch.org/pytorch/executorch/6276/runtime-python-api-reference.html

Reviewed By: dbort

Differential Revision: D64449105

Pulled By: larryliu0820

larryliu0820 commented:
@pytorchbot cherry-pick --onto release/0.4 -c docs
